src/Storages/MergeTree/KeyCondition.cpp

```diff
@@ -591,19 +592,36 @@ void KeyCondition::traverseAST(const ASTPtr & node, ContextPtr context, Block &
     rpn.emplace_back(std::move(element));
 }
 
-
 bool KeyCondition::canConstantBeWrappedByMonotonicFunctions(
     const ASTPtr & node,
     size_t & out_key_column_num,
     DataTypePtr & out_key_column_type,
     Field & out_value,
     DataTypePtr & out_type)
 {
-    // Constant expr should use alias names if any
-    String expr_name = node->getColumnName();
     const auto & sample_block = key_expr->getSampleBlock();
-    if (!sample_block.has(expr_name))
-        return false;
+
+    // Constant expr should use alias names if any
+    String passed_expr_name = node->getColumnName();
+    String expr_name;
+
+    /// sample_block from key_expr cannot contain modulo and moduloLegacy at the same time.
+    /// For partition key it is always moduloLegacy.
+    if (sample_block.has(passed_expr_name))
```

Comment:
Maybe move to a separate private KeyCondition method and add a huge `NOTE:` about backward compatibility and other explanation about this strange code.